inflate_p.h

```diff
@@ -211,3 +212,59 @@ static inline void window_output_flush(PREFIX3(stream) *strm) {
     state->whave += out_bytes;
     state->whave = MIN(state->whave, state->wsize);
 }
+
+/* Behave like chunkcopy, but avoid writing beyond of legal output. */
+static inline uint8_t* chunkcopy_safe(uint8_t *out, uint8_t *from, unsigned len, uint8_t *safe) {
+    unsigned safelen = (unsigned)((safe - out) + 1);
+    len = MIN(len, safelen);
+    int olap_src = ((from >= out && from < out + len) ||
+                    (from <= out && from + len > out));
+    int olap_dst = ((out >= from && out < from + len) || 
+                    (from <= out && out + len > from));
+
+    /* For all cases without overlap, memcpy is ideal */
```

Comment:
I am wondering how is it that `memcpy` isn't faster in all cases..

Reply:
I remember you were using `memcpy` before entirely, what happened to that?

Reply:
It's not so much that it's less than ideal performance wise (though, it might be), but correctness-wise it's not ideal to invoke memcpy when there's overlap.

Reply:
It's just a whole of branchiness and ugliness to handle the overlap to simulate a byte for byte copy.  I think there needs to be 2 or more branches depending on where the overlap lives.  It's tricky to get right but it's probably worth a separate pull request.  I just want this to work correctly and have the common case be fast for now.

Reply:
I think I have this resolved for now, anyway.  At least, we're no worse off now for robustness than before.  I'll have to wrap my head around a sort of "inverted memmove" that does the opposite intention of memmove.  Right now I think this is working by virtue of the compounded smaller copies at a time getting lucky enough that we still get the self modifying copy side effects.

I could, as a precaution, modify the overlapping copy code to do an explicit byte by byte copy loop (I think this happens infrequently enough that the performance penalty wouldn't be substantial).  The more optimal solution is solving for both directions of overlap with piecewise memcpy's though.